internal/testing/enginetest/enginetest.go

```diff
@@ -212,21 +215,188 @@ func RunTestModuleEngine_Call_HostFn(t *testing.T, et EngineTester) {
+			me:          importedMe,
+			fn:          imported.Exports[hostFnName].Function,
+			expectedErr: `expected 1 params, but passed 2`,
+		},
+		{
+			name:  "host function that panics",
+			input: []uint64{math.MaxUint32},
+			me:    importedMe,
+			fn:    imported.Exports[hostFnName].Function,
+			expectedErr: fmt.Sprintf(`host-function panic (recovered by wazero)
+wasm stack trace:
+	%[1]s.get_uint32(i32) i32`, t.Name()),
+		},
+		{
+			name:  "host function panics with runtime.Error",
+			input: []uint64{1},
+			me:    importedMe,
+			fn:    imported.Exports[hostFnName].Function,
+			expectedErr: fmt.Sprintf(`runtime error: invalid memory address or nil pointer dereference (recovered by wazero)
+wasm stack trace:
+	%[1]s.get_uint32(i32) i32`, t.Name()),
+		},
+		{
+			name:  "wasm calls host function that panics",
+			input: []uint64{math.MaxUint32},
+			me:    importedMe,
+			fn:    imported.Exports[callHostFnName].Function,
+			expectedErr: fmt.Sprintf(`host-function panic (recovered by wazero)
+wasm stack trace:
+	%[1]s.get_uint32(i32) i32
+	%[1]s.call__get_uint32(i32) i32`, t.Name()),
+		},
+		{
+			name:  "wasm calls imported wasm that calls host function panics with runtime.Error",
+			input: []uint64{1},
+			me:    importedMe,
+			fn:    imported.Exports[callHostFnName].Function,
+			expectedErr: fmt.Sprintf(`runtime error: invalid memory address or nil pointer dereference (recovered by wazero)
+wasm stack trace:
+	%[1]s.get_uint32(i32) i32
+	%[1]s.call__get_uint32(i32) i32`, t.Name()),
+		},
+		{
+			name:  "wasm calls imported wasm that calls host function that panics",
+			input: []uint64{math.MaxUint32},
+			me:    importingMe,
+			fn:    importing.Exports[callImportCallHostFnName].Function,
+			expectedErr: fmt.Sprintf(`host-function panic (recovered by wazero)
+wasm stack trace:
+	%[1]s.get_uint32(i32) i32
+	%[1]s.call__get_uint32(i32) i32
+	%[1]s_importing.call_import__call__get_uint32(i32) i32`, t.Name()),
+		},
+		{
+			name:  "wasm calls imported wasm calls host function panics with runtime.Error",
+			input: []uint64{1},
+			me:    importingMe,
+			fn:    importing.Exports[callImportCallHostFnName].Function,
+			expectedErr: fmt.Sprintf(`runtime error: invalid memory address or nil pointer dereference (recovered by wazero)
+wasm stack trace:
+	%[1]s.get_uint32(i32) i32
+	%[1]s.call__get_uint32(i32) i32
+	%[1]s_importing.call_import__call__get_uint32(i32) i32`, t.Name()),
+		},
+	}
+	for _, tt := range tests {
+		tc := tt
+
+		t.Run(tc.name, func(t *testing.T) {
+			ctx := newModuleContext(tc.fn.Module, tc.me)
+			_, err := tc.me.Call(ctx, tc.fn, tc.input...)
+			require.EqualError(t, err, tc.expectedErr)
+		})
+	}
+}
+
+const (
+	hostFnName               = "get_uint32"
+	callHostFnName           = "call->" + hostFnName
+	callImportCallHostFnName = "call_import->" + callHostFnName
+)
+
+func setupHostFunctions(t *testing.T, e wasm.Engine) (*wasm.ModuleInstance, wasm.ModuleEngine, *wasm.ModuleInstance, wasm.ModuleEngine) {
```

Comment:
ok at least you must set ModuleInstance.Engine to the created moduleEngine:
```diff
@@ -372,6 +372,8 @@ func setupHostFunctions(t *testing.T, e wasm.Engine) (*wasm.ModuleInstance, wasm
        importedMe, err := e.NewModuleEngine(imported.Name, nil, imported.Functions, nil, nil)
        require.NoError(t, err)

+       imported.Engine = importedMe
+
        // To test stack traces, call the same function from another module
        importing := &wasm.ModuleInstance{Name: t.Name() + "-importing"}
        callImportedHostFn := &wasm.FunctionInstance{
@@ -386,6 +388,8 @@ func setupHostFunctions(t *testing.T, e wasm.Engine) (*wasm.ModuleInstance, wasm
        importingMe, err := e.NewModuleEngine(importing.Name, []*wasm.FunctionInstance{callHostFn}, importing.Functions, nil, nil)
        require.NoError(t, err)

+       importing.Engine = importingMe
+
        return imported, importedMe, importing, importingMe
```

Reply:
as the native code emits under the assumption that the engine is not nil (e.g. amd64: https://github.com/tetratelabs/wazero/blob/main/internal/wasm/jit/jit_impl_amd64.go#L3975) which is guaranteed by the normal code path in store

Reply:
flak has gone at least on my local arm64/darwin `go test -timeout 10s -tags proxytest -run ^TestJIT_ModuleEngine_Call_HostFn$ github.com/tetratelabs/wazero/internal/wasm/jit -count=10000` passes with ^^ patch

Reply:
cool. we just need to know where all the special places are as otherwise a seemingly innocent refactor of store could end up with the same problem

Reply:
fwiw that was already done here:
```
func newModuleContext(module *wasm.ModuleInstance, engine wasm.ModuleEngine) *wasm.ModuleContext {
	// moduleInstanceEngineOffset
	module.Engine = engine
```

Reply:
yeah I agreed

Reply:
As far as I can tell for now, settings ModuleEngine is the only thing we must care on this.. (I hope..)

Reply:
> fwiw that was already done here:
> 
> ```
> func newModuleContext(module *wasm.ModuleInstance, engine wasm.ModuleEngine) *wasm.ModuleContext {
> 	// moduleInstanceEngineOffset
> 	module.Engine = engine
> ```

let's see..

Reply:
well actually this is enough to solve..
```diff
@@ -372,6 +372,8 @@ func setupHostFunctions(t *testing.T, e wasm.Engine) (*wasm.ModuleInstance, wasm
        importedMe, err := e.NewModuleEngine(imported.Name, nil, imported.Functions, nil, nil)
        require.NoError(t, err)

+       imported.Engine = importedMe
+
        // To test stack traces, call the same function from another module
        importing := &wasm.ModuleInstance{Name: t.Name() + "-importing"}
        callImportedHostFn := &wasm.FunctionInstance{
```

Reply:
oh yeah without this, the functions without ModuleEngine can be called right?

Reply:
`newModuleContext` is called inside the for-loop the order of `newModuleCotext` is random since the for-loop is over map, and if the importing module comes first, and makes call against the (yet-initialized) host-module, then this crash happens

Reply:
made this not in a loop in the other PR which this will rebase over.



```go
	me, err := e.NewModuleEngine(module.Name, nil, module.Functions, nil, nil)
	require.NoError(t, err)
	defer closeModuleEngineWithExitCode(t, me, 0)
	linkModuleToEngine(module, me)
--snip--

func linkModuleToEngine(module *wasm.ModuleInstance, me wasm.ModuleEngine) {
	module.Engine = me // for JIT, links the module to the module-engine compiled from it (moduleInstanceEngineOffset).
	// callEngineModuleContextModuleInstanceAddressOffset
	module.Ctx = wasm.NewModuleContext(context.Background(), nil, module, nil)
}
```